compiler/dyno/lib/uast/Builder.cpp

```diff
@@ -280,12 +294,136 @@ void Builder::doAssignIDs(AstNode* ast, UniqueString symbolPath, int& i,
+  void Builder::checkConfigPreviouslyUsed(const AstNode* ast, std::string& configNameUsed) {
+    // If you're reading this and confused about how we can call useConfigSetting
+    // and then call nameToConfigSetting, essentially setting a value and then
+    // asking for it back and comparing against the value we just used, you're not alone.
+    // See the docs in query-impl.h (QUERY_STORE_INPUT_RESULT) that describes
+    // why this works.
+    // "If called multiple times __within the same revision__, only the first
+    // stored result in that revision will be saved."
+    parsing::useConfigSetting(context(), configNameUsed, ast->id());
+    auto usedId = parsing::nameToConfigSettingId(context(), configNameUsed);
+
+    if (usedId != ast->id()) {
+      std::string err = "ambiguous config name (";
+      err += configNameUsed;
+      err += ")";
+      std::string otherLoc = "also defined here";
+      std::string disambiguate = "(disambiguate using -s<modulename>.";
+      disambiguate += configNameUsed;
+      disambiguate += "...)";
+      ErrorMessage errorMessage = ErrorMessage(ast->id(), idToLocation_[ast->id()], err, ErrorMessage::ERROR );
+      ErrorMessage noteOtherLoc = ErrorMessage(ast->id(), idToLocation_[usedId], otherLoc, ErrorMessage::NOTE);
+      ErrorMessage noteDisambiguate = ErrorMessage(ast->id(), idToLocation_[usedId], disambiguate, ErrorMessage::NOTE);
+      addError(errorMessage);
+      addError(noteOtherLoc);
+      addError(noteDisambiguate);
+    }
+  }
+
+  std::pair<std::string, std::string> Builder::nodeMatchesConfig(AstNode* ast, pathVecT& pathVec) {
+    std::pair<std::string, std::string> configMatched;
+    assert(ast->isVariable() && ast->toVariable()->isConfig());
+    auto var = ast->toVariable();
+    const auto &configs = parsing::configSettings(this->context());
+    // inspect pathVec
+    std::string possibleModule;
+    if (pathVec.size() > 1) {
+      for (size_t i = 1; i < pathVec.size(); i++) {
+        possibleModule += pathVec[i].first.str();
+        possibleModule += ".";
+      }
+    } else if (pathVec.size() == 1) {
+      possibleModule = pathVec[0].first.str();
+      possibleModule += ".";
+    }
+    // for config vars, check if they were set from the command line
+    for (auto configPair: configs) {
+      if ((var->name().str() == configPair.first && var->visibility() != Decl::PRIVATE)
+          || configPair.first == possibleModule + var->name().str()) {
+        // found a config that was set via cmd line
+        // handle deprecations
+        if (auto attribs = var->attributes()) {
+          if (attribs->isDeprecated()) {
+            // TODO: Need proper message handling here
+            std::string msg = "'" + var->name().str() + "' was set via a compiler flag";
+            if (attribs->deprecationMessage().isEmpty()) {
+              std::cerr << "warning: " + var->name().str() + " is deprecated" << std::endl;
```

Comment:
I think we should be emitting errors rather than printing to `cerr`.

Oh wait, that's what the TODO was for xD xD!

Reply:
Check out `noteError` and friends, Michael added some nice varargs overloads that kind of turn them into string builders (saves all the `std::string` concatenation pain).